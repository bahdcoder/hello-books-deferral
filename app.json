{
  "name": "Hello Books",
  "description": "A system that allows efficient, cost-effective management of Books for Libraries."
}
