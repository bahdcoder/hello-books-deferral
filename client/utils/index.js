export const getDateStr = str => {
    const end = str.indexOf('T');
    const rawDate = str.slice(0, end);
    const firstHyphen = rawDate.indexOf('-');
    const lastHyphen = rawDate.lastIndexOf('-');
    const year = rawDate.slice(0, firstHyphen);
    const month = rawDate.slice(firstHyphen + 1, lastHyphen);
    const day = rawDate.slice(lastHyphen + 1);
    const monthStrArr = [
        'Jan',
        'Feb',
        'Mar',
        'Apr',
        'May',
        'Jun',
        'Jul',
        'Aug',
        'Sep',
        'Oct',
        'Nov',
        'Dec'
    ];
    const monthStr = monthStrArr[Number(month) - 1];
    return `${monthStr} ${day} ${year}`;
};
