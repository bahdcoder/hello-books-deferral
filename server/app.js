import express from 'express';
import consola from 'consola';
import bodyParser from 'body-parser';
import swagger from 'swagger-ui-express';
import Knex from 'knex';
import { createServer } from 'http';
import { Model } from 'objection';
import swaggerDocument from '@/swagger.json';
import Routes from '@routes/v1';
import config from '@config';

import 'express-jsend';

export const app = express();

export const databaseConnection = Knex(config.database[config.server.env]);

Model.knex(databaseConnection);

app.use(bodyParser.json());
app.use(bodyParser.urlencoded({ extended: true }));

Routes(app);

app.get('/', (req, res) => {
    res.status(200).jsend({ message: 'Hello Books Deferral' });
});
/**
 *  setup Swagger
 */
app.use(
    '/api-docs',
    swagger.serve,
    swagger.setup(swaggerDocument, { explorer: true })
);

export const server = createServer(app);

export default { app, databaseConnection };
