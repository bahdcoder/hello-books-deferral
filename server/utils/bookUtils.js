const { validations } = require("indicative");
const { Vanilla } = require("indicative/builds/formatters");
const Validator = require("indicative/builds/validator");
const Book = require("../models/Book");

/* custom user friendly error messages */
const messages = {
  required: "{{ field }} is required to create a book",
  string: "{{ field }} is not a string",
  number: "{{ field }} is not a number",
  different: "{{ field }} must be unique"
};

// Trim white spaces
const sanitizeRules = {
  title: "trim",
  coverType: "trim",
  description: "trim",
  isbn: "trim",
  publisher: "trim",
  year: "trim",
  copiesAvailable: "trim"
};

/* add the unique custom validator to indicative validations object */
validations.different = async (data, field, message, args, get) => {
  const value = get(data, field);
  if (!value) return;
  const column = args[1];
  const row = await Book.query().where(column, value);
  if (row[0]) throw message;
};

const validatorInstance = Validator(
  validations,
  Vanilla
); /* create custom validator */

module.exports = { validatorInstance, messages, sanitizeRules };
