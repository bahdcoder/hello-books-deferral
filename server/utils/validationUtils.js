import { validations } from "indicative";
import { Vanilla } from "indicative/builds/formatters";
import Validator from "indicative/builds/validator";
import User from "@models/User";

/* custom user friendly error messages */
export const messages = {
  required: "{{ field }} is required to create a new account",
  string: "{{ field }} is not a string",
  unique: "{{ field }} must be unique",
  email: "{{ field }} is invalid",
  min: "{{ field }} is must be less than {{ argument.0 }}",
  alpha_numeric: "only numbers and letters are allowed for {{ field }}",
  confirmed: "Confirm {{ field }} does not match {{ field }}"
};

/* sanitization rule to trim whitespaces */
export const sanitizeRules = {
  firstName: "trim",
  lastName: "trim",
  email: "trim",
  password: "trim"
};

/* add the unique custom validator to indicative validations object */
validations.unique = async (data, field, message, args, get) => {
  const value = get(data, field);
  if (!value) return;
  const column = args[1];
  /* check user table to see if there is an existing user email */
  const row = await User.query().where(column, value);
  if (row[0]) throw message;
};

export const validatorInstance = Validator(
  validations,
  Vanilla
); /* create custom validator */
