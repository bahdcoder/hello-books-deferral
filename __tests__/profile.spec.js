import supertest from 'supertest';
import { app, databaseConnection } from '@server/app';
import { getUser, createUser, getToken } from '@tests/utils/helpers';

const server = () => supertest(app);
let userToken;
let userId;
let user;

describe('PROFILES API ENDPOINTS', () => {
    beforeAll(async () => {
        await databaseConnection('users').truncate();
        await databaseConnection.migrate.latest();
        user = await createUser(getUser());
        userToken = getToken({ id: user.id, email: user.email });
        const userB = await createUser(getUser());
        userId = userB.id;
    });

    afterAll(async () => {
        await databaseConnection('users').truncate();
        server.close();
    });

    describe('GET PROFILE api/v1/profile', () => {
        'should get profile for authenticated user',
            async () => {
                const { status, body } = await server()
                    .get('/api/v1/profile')
                    .set('x-access-token', userToken);
                expect(status).toBe(200);
                expect(Object.keys(body.data)).toMatchSnapshot();
            };

        it('should get profile for other users', async () => {
            const { status, body } = await server()
                .get(`/api/v1/profile/${userId}`)
                .set('x-access-token', userToken);

            expect(status).toBe(200);
            expect(Object.keys(body.data)).toMatchSnapshot();
        });

        it('should not get profile if userId param is not a number', async () => {
            const { status, body } = await server()
                .get('/api/v1/profile/yui')
                .set('x-access-token', userToken);
            expect(status).toBe(422);
            expect(body).toMatchSnapshot();
        });
    });

    describe('PATCH update profile api/v1/profile', () => {
        it('should not edit profile if all field is empty', async () => {
            const { status, body } = await server()
                .patch('/api/v1/profile')
                .set('x-access-token', userToken)
                .send({});

            expect(status).toBe(422);
            expect(body).toMatchSnapshot();
        });

        it('should not edit profile with an email that already exists', async () => {
            const { status, body } = await server()
                .patch('/api/v1/profile')
                .set('x-access-token', userToken)
                .send({
                    ...user,
                    bio: 'loves reading',
                    avatar: 'profilePic'
                });
            expect(status).toBe(422);
            expect(body).toMatchSnapshot();
        });

        it('should edit profile with valid inputs', async () => {
            const details = await getUser();
            const { status, body } = await server()
                .patch('/api/v1/profile')
                .set('x-access-token', userToken)
                .send({
                    ...details,
                    bio: 'i love football',
                    avatar: 'my pics'
                });

            expect(status).toBe(200);
            expect(Object.keys(body.data)).toMatchSnapshot();
        });
    });
});
