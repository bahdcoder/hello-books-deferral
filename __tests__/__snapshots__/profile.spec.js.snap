// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PROFILES API ENDPOINTS GET PROFILE api/v1/profile should get profile for other users 1`] = `
Array [
  "firstName",
  "lastName",
  "email",
]
`;

exports[`PROFILES API ENDPOINTS GET PROFILE api/v1/profile should not get profile if userId param is not a number 1`] = `
Object {
  "code": "ValidationFailed",
  "message": Array [
    Object {
      "field": "id",
      "message": "id is expected to be a an integer",
      "validation": "number",
    },
    Object {
      "field": "id",
      "message": "select \\"users\\".* from \\"users\\" where \\"id\\" = $1 - invalid input syntax for integer: \\"yui\\"",
      "validation": "ENGINE_EXCEPTION",
    },
  ],
  "status": "error",
}
`;

exports[`PROFILES API ENDPOINTS PATCH update profile api/v1/profile should not edit profile if all field is empty 1`] = `
Object {
  "code": Object {
    "message": "Update at least a field",
  },
  "status": "error",
}
`;

exports[`PROFILES API ENDPOINTS PATCH update profile api/v1/profile should not edit profile with an email that already exists 1`] = `
Object {
  "code": "ValidationFailed",
  "message": Array [
    Object {
      "field": "email",
      "message": "email must be unique",
      "validation": "unique",
    },
  ],
  "status": "error",
}
`;
